src/hotspot/share/gc/shared/stringdedup/stringDedupTable.cpp

```diff
@@ -395,31 +395,33 @@ bool StringDedupTable::is_rehashing() {
 StringDedupTable* StringDedupTable::prepare_resize() {
   size_t size = _table->_size;
 
-  // Check if the hashtable needs to be resized
+  // Decide whether to resize, and compute desired new size if so.
   if (_table->_entries > _table->_grow_threshold) {
-    // Grow table, double the size
-    size *= 2;
-    if (size > _max_size) {
-      // Too big, don't resize
-      return NULL;
+    // Compute new size.
+    size_t needed = _table->_entries / _grow_load_factor;
+    if (needed < _max_size) {
+      size = round_up_power_of_2(needed);
+    } else {
+      size = _max_size;
     }
   } else if (_table->_entries < _table->_shrink_threshold) {
-    // Shrink table, half the size
-    size /= 2;
-    if (size < _min_size) {
-      // Too small, don't resize
-      return NULL;
+    // Compute new size.
+    size_t needed = _table->_entries / _shrink_load_factor;
+    if (needed > _min_size) {
+      size = round_down_power_of_2(needed);
```

Comment:
Unfortunately, aggressive shrinking can't be done without other changes.
The current resizing code assumes shrinking is only by a factor of 2,
because of the way segments are claimed.  Shrinking could probably be
adjusted to deal with larger factors, but I think that's outside the scope
of what's needed for this change.  (Growth by an arbitrary power of 2
doesn't seem to be a problem.)  Fixing...

Reply:
I agree that it should be a separate change as long as the test failure is fixed. But I guess adjusting the grow should be enough to make the test pass, right?

Reply:
I think so, barring something pretty strange happening.

The initial table size is _min_size.  With the old growth calculation, if
the table gets big enough quickly enough then there will be enough growth
steps to occupy all of the collections and processing the test performs and
there won't be any rehashes, leading to test failure.  With the new
calculation, since the test program stops allocating a significant number of
strings before running the test GCs, the size will jump and then remain
stable for some rehashes to get in.  If a lot of deduped strings were
dropped so that many shrink steps were needed, that could break the test,
but the test doesn't do that, and it seems unlikely that would happen for
environmental or other reasons.